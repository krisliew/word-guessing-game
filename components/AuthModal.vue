<template>
<div class="modal" :class="{'is-active': active}">
  <div class="modal-background" @click="cancel()"></div>
  <div class="modal-card">
    <header class="modal-card-head">
      <p class="modal-card-title">{{ title }}</p>
      <button class="delete" :disabled="loading" aria-label="close" @click="cancel()"></button>
    </header>
    <section class="modal-card-body">
      <div class="field">
        <label class="label">Username</label>
        <div class="control">
          <input v-model="username" class="input" type="text" placeholder="Username">
        </div>
      </div>
      <div class="field">
        <label class="label">Password</label>
        <div class="control">
          <input v-model="password" class="input" type="password" placeholder="Password">
        </div>
      </div>
      <article v-if="errorMessage" class="message is-danger">
        <div class="message-body">
          {{ errorMessage }}
        </div>
      </article>
    </section>
    <footer class="modal-card-foot">
      <button class="button is-success" :class="{'is-loading': loading}" @click="submit()">{{ submitText }}</button>
      <button class="button" :disabled="loading" @click="cancel()">{{ cancelText }}</button>
    </footer>
  </div>
</div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      required: true
    },
    submitText: {
      type: String,
      required: true
    },
    cancelText: {
      type: String,
      default: "Cancel"
    },
    active: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    errorMessage: {
      type: String,
      default: null
    }
  },
  data: () => {
    return {
      username: "",
      password: ""
    };
  },
  methods: {
    submit() {
      this.$emit("submit", {
        username: this.username,
        password: this.password
      });
    },
    cancel() {
      this.$emit("cancel");
    }
  }
};
</script>

<style scoped>
</style>
