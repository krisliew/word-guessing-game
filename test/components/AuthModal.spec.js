import AuthModal from "@/components/AuthModal.vue";
import { makeVueMock } from "@/test/utils";

describe("SignUpModal", () => {
  test("is vue instance", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit"
      }
    });

    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  /**
   * Snapshots
   */

  test("renders regular", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit"
      }
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("renders active", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit",
        active: true
      }
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("custom cancel text", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit",
        cancelText: "Don't Submit"
      }
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("loading", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit",
        cancelText: "Don't Submit",
        loading: true
      }
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("error message", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit",
        cancelText: "Don't Submit",
        errorMessage: "This is an error"
      }
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  /**
   * Event Testing
   */

  test("submit clicked", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit"
      }
    });

    wrapper.find(".modal-card-foot > button:first-child").trigger("click");

    const submitEmission = wrapper.emitted().submit;

    expect(submitEmission).toBeTruthy();
  });

  test("submit event with data", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit"
      }
    });

    wrapper.vm.$data.username = "Panda Bear";
    wrapper.vm.$data.password = "bamboo123";

    wrapper.find(".modal-card-foot > button:first-child").trigger("click");

    const submitEmission = wrapper.emitted().submit;

    expect(submitEmission[0]).toEqual([
      { username: "Panda Bear", password: "bamboo123" }
    ]);
  });

  test("cancel event - 'x' button", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit"
      }
    });

    wrapper.find(".modal-card-head > button.delete").trigger("click");

    const submitEmission = wrapper.emitted().cancel;

    expect(submitEmission).toBeTruthy();
  });

  test("cancel event - cancel button", () => {
    const { wrapper } = makeVueMock(AuthModal, {
      propsData: {
        title: "Auth Modal",
        submitText: "Submit"
      }
    });

    wrapper.find(".modal-card-foot > button:last-child").trigger("click");

    const submitEmission = wrapper.emitted().cancel;

    expect(submitEmission).toBeTruthy();
  });
});
