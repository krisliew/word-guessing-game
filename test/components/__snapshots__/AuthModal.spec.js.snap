// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpModal custom cancel text 1`] = `
<div
  class="modal"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Auth Modal
      </p>
       
      <button
        aria-label="close"
        class="delete"
      />
    </header>
     
    <section
      class="modal-card-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Username
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Username"
            type="text"
          />
        </div>
      </div>
       
      <div
        class="field"
      >
        <label
          class="label"
        >
          Password
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Password"
            type="password"
          />
        </div>
      </div>
       
      <!---->
    </section>
     
    <footer
      class="modal-card-foot"
    >
      <button
        class="button is-success"
      >
        Submit
      </button>
       
      <button
        class="button"
      >
        Don't Submit
      </button>
    </footer>
  </div>
</div>
`;

exports[`SignUpModal error message 1`] = `
<div
  class="modal"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Auth Modal
      </p>
       
      <button
        aria-label="close"
        class="delete"
      />
    </header>
     
    <section
      class="modal-card-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Username
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Username"
            type="text"
          />
        </div>
      </div>
       
      <div
        class="field"
      >
        <label
          class="label"
        >
          Password
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Password"
            type="password"
          />
        </div>
      </div>
       
      <article
        class="message is-danger"
      >
        <div
          class="message-body"
        >
          
          This is an error
        
        </div>
      </article>
    </section>
     
    <footer
      class="modal-card-foot"
    >
      <button
        class="button is-success"
      >
        Submit
      </button>
       
      <button
        class="button"
      >
        Don't Submit
      </button>
    </footer>
  </div>
</div>
`;

exports[`SignUpModal loading 1`] = `
<div
  class="modal"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Auth Modal
      </p>
       
      <button
        aria-label="close"
        class="delete"
        disabled="disabled"
      />
    </header>
     
    <section
      class="modal-card-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Username
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Username"
            type="text"
          />
        </div>
      </div>
       
      <div
        class="field"
      >
        <label
          class="label"
        >
          Password
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Password"
            type="password"
          />
        </div>
      </div>
       
      <!---->
    </section>
     
    <footer
      class="modal-card-foot"
    >
      <button
        class="button is-success is-loading"
      >
        Submit
      </button>
       
      <button
        class="button"
        disabled="disabled"
      >
        Don't Submit
      </button>
    </footer>
  </div>
</div>
`;

exports[`SignUpModal renders active 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Auth Modal
      </p>
       
      <button
        aria-label="close"
        class="delete"
      />
    </header>
     
    <section
      class="modal-card-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Username
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Username"
            type="text"
          />
        </div>
      </div>
       
      <div
        class="field"
      >
        <label
          class="label"
        >
          Password
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Password"
            type="password"
          />
        </div>
      </div>
       
      <!---->
    </section>
     
    <footer
      class="modal-card-foot"
    >
      <button
        class="button is-success"
      >
        Submit
      </button>
       
      <button
        class="button"
      >
        Cancel
      </button>
    </footer>
  </div>
</div>
`;

exports[`SignUpModal renders regular 1`] = `
<div
  class="modal"
>
  <div
    class="modal-background"
  />
   
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Auth Modal
      </p>
       
      <button
        aria-label="close"
        class="delete"
      />
    </header>
     
    <section
      class="modal-card-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Username
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Username"
            type="text"
          />
        </div>
      </div>
       
      <div
        class="field"
      >
        <label
          class="label"
        >
          Password
        </label>
         
        <div
          class="control"
        >
          <input
            class="input"
            placeholder="Password"
            type="password"
          />
        </div>
      </div>
       
      <!---->
    </section>
     
    <footer
      class="modal-card-foot"
    >
      <button
        class="button is-success"
      >
        Submit
      </button>
       
      <button
        class="button"
      >
        Cancel
      </button>
    </footer>
  </div>
</div>
`;
